test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,338 @@
+import { expect } from 'chai';
+import { EventEmitter } from 'events';
+import * as util from 'util';
+
+import { Connection, HostAddress } from '../../src';
+import { ConnectionPool } from '../../src/cmap/connection_pool';
+import { sleep } from './utils';
+
+type CmapOperation =
+  | { name: 'start' | 'waitForThread'; target: string }
+  | { name: 'wait'; ms: number }
+  | { name: 'waitForEvent'; event: string; count: number }
+  | { name: 'checkOut'; thread: string; label: string }
+  | { name: 'checkIn'; connection: string }
+  | { name: 'clear' | 'close' | 'ready' };
+
+type CmapPoolOptions = {
+  maxPoolSize?: number;
+  minPoolSize?: number;
+  maxIdleTimeMS?: number;
+  waitQueueTimeoutMS?: number;
+};
+
+type CmapEvent = {
+  type: string;
+  address?: 42;
+  connectionId?: number;
+  options?: 42 | CmapPoolOptions;
+  reason: string;
+};
+
+const knownTestKeys: Array<keyof CmapTest> = [
+  'name',
+  'version',
+  'style',
+  'description',
+  'poolOptions',
+  'operations',
+  'error',
+  'events',
+  'ignore'
+];
+export type CmapTest = {
+  name?: string; // filename path added by the spec loader
+  version: number;
+  style: 'unit';
+  description: string;
+  poolOptions?: CmapPoolOptions;
+  operations: CmapOperation[];
+  error?: {
+    type: string;
+    message: string;
+    address?: number;
+  };
+  events?: CmapEvent[];
+  ignore?: string[];
+};
+
+const ALL_POOL_EVENTS = new Set([
+  ConnectionPool.CONNECTION_POOL_CREATED,
+  ConnectionPool.CONNECTION_POOL_CLOSED,
+  ConnectionPool.CONNECTION_POOL_CLEARED,
+  ConnectionPool.CONNECTION_CREATED,
+  ConnectionPool.CONNECTION_READY,
+  ConnectionPool.CONNECTION_CLOSED,
+  ConnectionPool.CONNECTION_CHECK_OUT_STARTED,
+  ConnectionPool.CONNECTION_CHECK_OUT_FAILED,
+  ConnectionPool.CONNECTION_CHECKED_OUT,
+  ConnectionPool.CONNECTION_CHECKED_IN
+]);
+
+function getEventType(event) {
+  const eventName = event.constructor.name;
+  return eventName.substring(0, eventName.lastIndexOf('Event'));
+}
+
+/**
+ * In the cmap spec and runner definition,
+ * a "thread" refers to a concurrent execution context
+ */
+class Thread {
+  _promise: Promise<void>;
+  _error: Error;
+  _killed = false;
+
+  _knownCommands: any;
+
+  // concurrent execution context
+  constructor(operations) {
+    this._promise = new Promise(resolve => {
+      this.start = () => resolve();
+    });
+
+    this._knownCommands = operations;
+  }
+
+  start: () => void;
```

Comment:
nit: can we move this above the constructor so property declarations are grouped together?

Reply:
sure

Reply:
done